examples/shader/custom_render_phase.rs

```diff
@@ -0,0 +1,630 @@
+        },
+        renderer::RenderContext,
+        sync_world::MainEntity,
+        view::{ExtractedView, RenderVisibleEntities, RetainedViewEntity, ViewTarget},
+        Extract, Render, RenderApp, RenderSet,
+    },
+};
+use nonmax::NonMaxU32;
+
+const SHADER_ASSET_PATH: &str = "shaders/custom_stencil.wgsl";
+
+fn main() {
+    App::new()
+        .add_plugins((DefaultPlugins, MeshStencilPhasePlugin))
+        .add_systems(Startup, setup)
+        .run();
+}
+
+fn setup(
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+) {
+    // circular base
+    commands.spawn((
+        Mesh3d(meshes.add(Circle::new(4.0))),
+        MeshMaterial3d(materials.add(Color::WHITE)),
+        Transform::from_rotation(Quat::from_rotation_x(-std::f32::consts::FRAC_PI_2)),
+    ));
+    // cube
+    // This cube will be rendered by the main pass, but it will also be rendered by our custom
+    // pass. This should result in an unlit red cube
+    commands.spawn((
+        Mesh3d(meshes.add(Cuboid::new(1.0, 1.0, 1.0))),
+        MeshMaterial3d(materials.add(Color::srgb_u8(124, 144, 255))),
+        Transform::from_xyz(0.0, 0.5, 0.0),
+        // This marker component is used to identify which mesh will be used in our custom pass
+        // The circle doesn't have it so it won't be rendered in our pass
+        DrawStencil,
+    ));
+    // light
+    commands.spawn((
+        PointLight {
+            shadows_enabled: true,
+            ..default()
+        },
+        Transform::from_xyz(4.0, 8.0, 4.0),
+    ));
+    // camera
+    commands.spawn((
+        Camera3d::default(),
+        Transform::from_xyz(-2.0, 4.5, 9.0).looking_at(Vec3::ZERO, Vec3::Y),
+        // disable msaa for simplicity
+        Msaa::Off,
+    ));
+}
+
+#[derive(Component, ExtractComponent, Clone, Copy, Default)]
+struct DrawStencil;
+
+struct MeshStencilPhasePlugin;
+impl Plugin for MeshStencilPhasePlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins((ExtractComponentPlugin::<DrawStencil>::default(),));
+        // We need to get the render app from the main app
+        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+        render_app
+            .init_resource::<SpecializedMeshPipelines<StencilPipeline>>()
+            .init_resource::<DrawFunctions<StencilPhase>>()
+            .add_render_command::<StencilPhase, DrawMesh3dStencil>()
+            .init_resource::<ViewSortedRenderPhases<StencilPhase>>()
+            .add_systems(ExtractSchedule, extract_camera_phases)
+            .add_systems(
+                Render,
+                (
+                    sort_phase_system::<StencilPhase>.in_set(RenderSet::PhaseSort),
+                    batch_and_prepare_sorted_render_phase::<StencilPhase, StencilPipeline>
+                        .in_set(RenderSet::PrepareResources),
+                    queue_custom_meshes.in_set(RenderSet::QueueMeshes),
```

Comment:
I would put the systems in a different order, as I was a bit confused when reading this; it seems like the order of execution is actually
```
// queue entities with DrawStencil as PhaseItems for the new render phase
queue_custom_meshes.in_set(RenderSet::QueueMeshes),
// sort
sort_phase_system::<StencilPhase>.in_set(RenderSet::PhaseSort),
// batch
batch_and_prepare_sorted_render_phase::<StencilPhase, StencilPipeline>
                        .in_set(RenderSet::PrepareResources),
```

Reply:
I just wrote them in the order I originally wrote the systems. I could maybe try ordering everything based on the execution order.